File: app/src/main/java/me/ykrank/s1next/view/adapter/simple/SimpleAdapterDelegate.java
package me.ykrank.s1next.view.adapter.simple;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.hannesdorfmann.adapterdelegates3.AdapterDelegate;

import java.util.List;

import me.ykrank.s1next.BR;
import me.ykrank.s1next.view.adapter.item.ProgressItem;

public class SimpleAdapterDelegate extends AdapterDelegate<List<Object>> {
    private final LayoutInflater mLayoutInflater;
    private final int layoutRes;
    @Nullable
    private final BindViewHolderCallback bindViewHolderCallback;

    SimpleAdapterDelegate(Context context, @LayoutRes int layoutRes, @Nullable BindViewHolderCallback bindViewHolderCallback) {
        mLayoutInflater = LayoutInflater.from(context);
        this.layoutRes = layoutRes;
        this.bindViewHolderCallback = bindViewHolderCallback;
    }

    @Override
    protected boolean isForViewType(@NonNull List<Object> items, int position) {
        return !ProgressItem.class.isInstance(items.get(position));
    }

    @NonNull
    @Override
    protected RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent) {
        ViewDataBinding binding = DataBindingUtil.inflate(mLayoutInflater, layoutRes, parent, false);
        return new SimpleRecycleViewHolder<>(binding);
    }

    @Override
    protected void onBindViewHolder(@NonNull List<Object> items, int position, @NonNull RecyclerView.ViewHolder holder, @NonNull List<Object> payloads) {
        ViewDataBinding binding = ((SimpleRecycleViewHolder) holder).binding;
        binding.setVariable(BR.model, items.get(position));
        if (bindViewHolderCallback != null) {
            bindViewHolderCallback.onBindViewHolder(binding);
        }
        binding.executePendingBindings();
    }
}

File: library/src/main/java/com/github/ykrank/androidtools/ui/adapter/simple/SimpleAdapterDelegate.kt
package com.github.ykrank.androidtools.ui.adapter.simple

import android.content.Context
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.github.ykrank.androidtools.GlobalData
import com.github.ykrank.androidtools.ui.adapter.delegate.item.FooterProgressItem
import com.github.ykrank.androidtools.ui.adapter.delegate.item.ProgressItem
import com.hannesdorfmann.adapterdelegates3.AdapterDelegate

open class SimpleAdapterDelegate constructor(context: Context, @param:LayoutRes private val layoutRes: Int,
                                             private val modelClass: Class<*>? = null,
                                             private val createViewHolderCallback: ((ViewDataBinding) -> Unit)? = null,
                                             private val bindViewHolderCallback: BindViewHolderCallback? = null) : AdapterDelegate<MutableList<Any>>() {
    private val mLayoutInflater: LayoutInflater = LayoutInflater.from(context)

    override fun isForViewType(items: MutableList<Any>, position: Int): Boolean {
        val item = items[position]
        if (modelClass == null) {
            return !ProgressItem::class.java.isInstance(item) && !FooterProgressItem::class.java.isInstance(item)
        } else {
            return modelClass.isInstance(item)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(mLayoutInflater, layoutRes, parent, false)
        createViewHolderCallback?.invoke(binding)
        return SimpleRecycleViewHolder(binding)
    }

    override fun onBindViewHolder(items: MutableList<Any>, position: Int, holder: androidx.recyclerview.widget.RecyclerView.ViewHolder, payloads: List<Any>) {
        val binding = (holder as SimpleRecycleViewHolder<*>).binding
        binding.setVariable(GlobalData.provider.itemModelBRid, items[position])
        bindViewHolderCallback?.onBindViewHolder(position, binding)
    }
}